test/latency/latency-part1.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2017 Intel Corporation.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/intel-go/yanff/flow"
+	"github.com/intel-go/yanff/packet"
+)
+
+// Latency test includes 2 parts:
+//
+// latency-part1:
+// This part of test generates packets, puts time of generation into packet
+// and send to port 0. Packets expected to be received back on 1 port. Received
+// flow is partitioned in 100000:1 relation. For this 1 packet latency is
+// calculated (current time minus time in packet) and then reported to channel.
+// Test is finished when number of reported latency measurements = LAT_NUMBER
+//
+// latency-part2:
+// This part of test receives packets on 0 port and send it back to 1 port.
+
+const (
+	// LAT_NUMBER is number of packets, for which latency should be reported.
+	LAT_NUMBER = 500
+	BINS       = 10
+)
+
+var (
+	// PAYLOAD_SIZE is size of PacketData struct
+	// on amd64. PacketData keeps packet send time.
+	PAYLOAD_SIZE uint64 = 24
+
+	// Counters of sent packets
+	sentPackets uint64 = 0
+	// Counter of packets, that were checked with handler
+	checkedPackets uint64 = 0
+
+	// Event is to notify when LAT_NUMBER number of packets are received.
+	testDoneEvent *sync.Cond = nil
+
+	// Channel is used to report packet latencies
+	latencies chan time.Duration
+	// Channel to stop collecting latencies and print statistics
+	stop chan string
+	// Latency values are stored here for next processing
+	latenciesStorage []time.Duration
+)
+
+func main() {
+	latencies = make(chan time.Duration)
+	stop = make(chan string)
+	latenciesStorage = make([]time.Duration, 2*LAT_NUMBER)
+
+	go latenciesLogger(latencies, stop)
+
+	// Event needed to stop sending packets
+	var m sync.Mutex
+	testDoneEvent = sync.NewCond(&m)
+
+	// Initialize YANFF library at 16 available cores
+	flow.SystemInit(16)
+
+	// Create packet flow
+	outputFlow := flow.SetGenerator(generatePackets, 10000)
+	flow.SetSender(outputFlow, 0)
+
+	// Create receiving flow and set a checking function for it
+	inputFlow := flow.SetReceiver(1)
+
+	// Calculate latency only for 1 of 100000 packets. Otherwise channel is conjested
+	latFlow := flow.SetPartitioner(inputFlow, 100000, 1)
+	flow.SetHandler(latFlow, checkPackets)
+
+	flow.SetStopper(inputFlow)
+	flow.SetStopper(latFlow)
+
+	// Start pipeline
+	go flow.SystemStart()
+
+	// Wait until enough latencies calculated
+	testDoneEvent.L.Lock()
+	testDoneEvent.Wait()
+	testDoneEvent.L.Unlock()
+
+	// Compose statistics
+	sent := atomic.LoadUint64(&sentPackets)
+	checked := atomic.LoadUint64(&checkedPackets)
+
+	// Leave time for goroutine to print latencies info to screen
+	time.Sleep(time.Second)
```

Comment:
Using Sleep is bad design. It is much better to use events or channels for such purpose. You are using testDoneEvent a few lines above. Sending such event could be done in line 128 of latenciesLogger.